packages/transaction-controller/src/helpers/ResimulateHelper.ts

```diff
@@ -1,31 +1,139 @@
+  updateSimulationData: (transactionMeta: TransactionMeta) => Promise<void>;
+};
+
+export class ResimulateHelper {
+  // Map of transactionId <=> isActive
+  readonly #activeResimulations: Map<string, boolean> = new Map();
+
+  // Map of transactionId <=> intervalId
+  readonly #intervalIds: Map<string, NodeJS.Timeout> = new Map();
+
+  readonly #getTransactions: () => TransactionMeta[];
+
+  readonly #updateSimulationData: (transactionMeta: TransactionMeta) => Promise<void>;
+
+  constructor({
+    getTransactions,
+    updateSimulationData,
+    onStateChange,
+  }: ResimulateHelperOptions) {
+    this.#getTransactions = getTransactions;
+    this.#updateSimulationData = updateSimulationData;
+
+    onStateChange(() => {
+      const unapprovedTransactions = this.#getUnapprovedTransactions();
+      const unapprovedTransactionIds = new Set(
+        unapprovedTransactions.map((tx) => tx.id),
+      );
+
+      // Start or stop resimulation based on the current isFocused state
+      unapprovedTransactions.forEach((transactionMeta) => {
+        if (transactionMeta.isFocused) {
+          this.#start(transactionMeta);
+        } else {
+          this.#stop(transactionMeta);
+        }
+      });
+
+      // Force stop any running active resimulation that are no longer unapproved transactions list
+      this.#activeResimulations.forEach((isActive, transactionId) => {
+        if (isActive && !unapprovedTransactionIds.has(transactionId)) {
+          this.#stop({
+            id: transactionId,
+            // Forcing this to false to ensure the resimulation is stopped
+            isFocused: false,
+          } as unknown as TransactionMeta);
+        }
+      });
+    });
+  }
+
+  #start(transactionMeta: TransactionMeta) {
+    const { id: transactionId } = transactionMeta;
+    const isActive = this.#activeResimulations.get(transactionId);
+    if (!transactionMeta.isFocused || isActive) {
+      return;
+    }
+
+    const listener = async () => {
+      try {
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        this.#updateSimulationData(transactionMeta);
+      } catch (error) {
+        /* istanbul ignore next */
+        log('Error during transaction resimulation', error);
+      }
+    };
+
+    // eslint-disable-next-line @typescript-eslint/no-misused-promises
+    const intervalId = setInterval(listener, RESIMULATE_INTERVAL_MS);
+
+    this.#intervalIds.set(transactionId, intervalId);
+    this.#activeResimulations.set(transactionId, true);
+    log(`Started resimulating transaction ${transactionId} every 3 seconds`);
+  }
+
+  #stop(transactionMeta: TransactionMeta) {
+    const { id: transactionId } = transactionMeta;
+    const isActive = this.#activeResimulations.get(transactionId);
+    if (transactionMeta.isFocused || !isActive) {
+      return;
+    }
+
+    this.#removeListener(transactionId);
+    log(`Stopped resimulating transaction ${transactionId} every 3 seconds`);
```

Comment:
Minor, should we avoid hardcoding the duration here so it's dynamic based on the constant?